pika/channel.py

```diff
@@ -1267,9 +1269,36 @@ def _on_synchronous_complete(self, _method_frame_unused):
         """
         LOGGER.debug('%i blocked frames', len(self._blocked))
         self._blocking = None
+        # self._blocking must be checked here as a callback could
+        # potentially change the state of that variable during an
+        # iteration of the while loop
         while self._blocked and self._blocking is None:
             self._rpc(*self._blocked.popleft())
 
+    def _drain_blocked_methods_on_remote_close(self):
+        """This is called when the broker sends a Channel.Close. It checks the
+        blocked method queue for pending client-initiated Channel.Close methods
+        and ensures their callbacks are processed
+
+        """
+        LOGGER.debug('Draining %i blocked frames due to remote Channel.Close', len(self._blocked))
+        self._blocking = None
+        # self._blocking must be checked here as a callback could
+        # potentially change the state of that variable during an
+        # iteration of the while loop
+        while self._blocked and self._blocking is None:
+            method = self._blocked.popleft()[0]
+            if isinstance(method, spec.Channel.Close):
```

Comment:
There should never be more than one `spec.Channel.Close`  in `self._blocked` because `Channel.close` sets the `CLOSING` state and will raise an exception if called again in that state. So can probably break out of the loop immediately after calling `self._on_close_meta`. Might be good to include an explanation to this effect as code commend before breaking out.

Reply:
What do you think about continuing the loop just so that any other blocked methods are debug-logged?

Reply:
> What do you think about continuing the loop just so that any other blocked methods are debug-logged?

May be marginal, because: 1. if channel errors out, the developer should reasonably expect that some initiated requests would not complete (also since they didn't get completion callbacks); and 2. The one(s) that made it into the transport buffer and/or socket would not show up in these debug statements.